src/graph/LookupExecutor.cpp

```diff
@@ -0,0 +1,499 @@
+            }
+            returnCols_.emplace_back(aExpr->prop()->c_str());
+        }
+    }
+    return Status::OK();
+}
+
+Status LookupExecutor::optimize() {
+    Status status = Status::OK();
+    do {
+        status = checkFilter();
+        if (!status.ok()) {
+            break;
+        }
+        status = chooseIndex();
+        if (!status.ok()) {
+            break;
+        }
+    } while (false);
+
+    if (!status.ok()) {
+        LOG(ERROR) << "Preparing failed: " << status;
+        return status;
+    }
+    return status;
+}
+
+Status LookupExecutor::traversalExpr(const Expression *expr) {
+    switch (expr->kind()) {
+        case nebula::Expression::kLogical : {
+            auto* lExpr = dynamic_cast<const LogicalExpression*>(expr);
+            if (lExpr->op() == LogicalExpression::Operator::XOR) {
+                return Status::SyntaxError("Unsupported syntax ：%s", lExpr->toString().c_str());
+            } else if (lExpr->op() == LogicalExpression::Operator::OR) {
+                skipOptimize_ = true;
+            }
+            auto* left = lExpr->left();
+            traversalExpr(left);
+            auto* right = lExpr->right();
+            traversalExpr(right);
+            break;
+        }
+        case nebula::Expression::kRelational : {
+            std::string prop;
+            VariantType v;
+            auto* rExpr = dynamic_cast<const RelationalExpression*>(expr);
+            auto* left = rExpr->left();
+            auto* right = rExpr->right();
+            if (left->kind() == nebula::Expression::kAliasProp
+                && right->kind() == nebula::Expression::kPrimary) {
+                auto* aExpr = dynamic_cast<const AliasPropertyExpression*>(left);
+                prop = *aExpr->prop();
+                filters_.emplace_back(std::make_pair(prop, rExpr->op()));
+            } else if (left->kind() == nebula::Expression::kPrimary
+                       && right->kind() == nebula::Expression::kAliasProp) {
+                auto* aExpr = dynamic_cast<const AliasPropertyExpression*>(right);
+                prop = *aExpr->prop();
+                filters_.emplace_back(std::make_pair(prop, rExpr->op()));
+            } else {
+                return Status::SyntaxError("Unsupported syntax ：%s", rExpr->toString().c_str());
+            }
+            break;
+        }
+        case nebula::Expression::kFunctionCall : {
+            return Status::SyntaxError("Unsupported function call ： %s",
+                                       expr->toString().c_str());
+        }
+        default : {
+            return Status::SyntaxError("Unsupported syntax ： %s", expr->toString().c_str());
+        }
+    }
+    return Status::OK();
+}
+
+Status LookupExecutor::checkFilter() {
+    auto status = traversalExpr(whereWrapper_->getFilter());
+    return status;
+}
+
+Status LookupExecutor::chooseIndex() {
+    auto ret = (isEdge_) ? ectx()->getMetaClient()->getEdgeIndexesFromCache(spaceId_)
```

Comment:
Put it in prepareClause.

Reply:
The purpose of `chooseIndex` is to select an optimal index, It required some conditions after the expression has been analyzed. I think it is a part of optimize, so I think it should be right here.

Reply:
I mean put the codes from line 198 to 208 into prepare step.